crates/sui/src/upgrade_compatibility.rs

```diff
@@ -0,0 +1,414 @@
+                old_enum: old_enum.clone(),
+                new_enum: new_enum.clone(),
+                tag: variant_idx,
+            });
+    }
+
+    fn function_missing_public(&mut self, name: &Identifier, old_function: &Function) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FunctionMissingPublic {
+                name: name.clone(),
+                old_function: old_function.clone(),
+            });
+    }
+
+    fn function_missing_friend(&mut self, name: &Identifier, old_function: &Function) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FunctionMissingFriend {
+                name: name.clone(),
+                old_function: old_function.clone(),
+            });
+    }
+
+    fn function_missing_entry(&mut self, name: &Identifier, old_function: &Function) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FunctionMissingEntry {
+                name: name.clone(),
+                old_function: old_function.clone(),
+            });
+    }
+
+    fn function_signature_mismatch(
+        &mut self,
+        name: &Identifier,
+        old_function: &Function,
+        new_function: &Function,
+    ) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FunctionSignatureMismatch {
+                name: name.clone(),
+                old_function: old_function.clone(),
+                new_function: new_function.clone(),
+            });
+    }
+
+    fn function_lost_public_visibility(&mut self, name: &Identifier, old_function: &Function) {
+        self.errors.push(
+            UpgradeCompatibilityModeError::FunctionLostPublicVisibility {
+                name: name.clone(),
+                old_function: old_function.clone(),
+            },
+        );
+    }
+
+    fn function_lost_friend_visibility(&mut self, name: &Identifier, old_function: &Function) {
+        self.errors.push(
+            UpgradeCompatibilityModeError::FunctionLostFriendVisibility {
+                name: name.clone(),
+                old_function: old_function.clone(),
+            },
+        );
+    }
+
+    fn function_entry_compatibility(
+        &mut self,
+        name: &Identifier,
+        old_function: &Function,
+        new_function: &Function,
+    ) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FunctionEntryCompatibility {
+                name: name.clone(),
+                old_function: old_function.clone(),
+                new_function: new_function.clone(),
+            });
+    }
+
+    fn friend_module_missing(
+        &mut self,
+        old_modules: BTreeSet<ModuleId>,
+        new_modules: BTreeSet<ModuleId>,
+    ) {
+        self.errors
+            .push(UpgradeCompatibilityModeError::FriendModuleMissing(
+                old_modules.clone(),
+                new_modules.clone(),
+            ));
+    }
+
+    fn finish(&self, _: &Compatibility) -> Result<(), Self::Error> {
```

Comment:
You will need to pay attention to the `Compatibility` flags here, when deciding which errors to signal to people.